### PR TITLE
Make it possible to override document within the app

### PR DESCRIPTION
**Pull request by allantatter**
Tarvo, Kaspar and other have had problems with create-frontend because they have needed to override the document HTML and decorate the rendered React app on the server-side. This should be a solution to these problems.

### CODE REVIEW

**Comment by allantatter** [comment]
The document can be overridden like this: https://d.pr/i/z14zUm, https://d.pr/i/kAMC7S

File: universal-react/server/wrapInDocument.js [hunk 1]
```diff
@@ -25,7 +25,17 @@ function getManifest() {
...
 }
 
-export default function wrapInDocument(content, appData, helmetContext, cspNonce, inlineStyles = []) {
+export default function wrapInDocument(
+  content,
+  appData,
+  helmetContext,
+  cspNonce,
+  inlineStyles = [],
+  overriddenDocument
```
Comment (KasparRosin) [inline universal-react/server/wrapInDocument.js:34]: I think you can do here `document = defaultDocument` as it's done for inline styles.
You can remove the L37 then.

File: universal-react/server/render.js [hunk 2]
```diff
@@ -15,7 +15,7 @@ import urlParser from 'url';
  *
  * @return {{ content: String, context: Object }}
  */
-export default async function renderOnServer(ReactComponent, url, props = {}, cspNonce) {
+export default async function renderOnServer(ReactComponent, url, props = {}, { cspNonce, appDecorator, document }) {
```
Comment (MatisLepik) [inline universal-react/server/render.js:18]: This should be updated in `docs/universal-react.md` too

**Comment by MatisLepik** [comment]
Document change looks good, although the API is getting really complicated. Might be worth rethinking it. Or at least, could the third and forth arguments be combined together into some `config` parameter?

For `appDecorator`: not sure what the use-case is, but couldn't it be solved without an API change?

```
// server/entry.js

function CustomApp(props) {
  // Custom stuff here
  return <App {...props} />;
}

const { content, context } = await render(CustomApp, ...);
```

**Comment by allantatter** [comment]
@MatisLepik The thing we'd like to solve with `appDecorator` is for example this: https://d.pr/i/JBIc7G
When we wrap the `<App />` with a custom component, then I'm not sure we can accomplish this. @KasparRosin what do you think?

**Comment by allantatter** [comment]
@MatisLepik which API is getting too compicated? `renderOnServer()`?

```
export default async function renderOnServer(ReactComponent, url, props = {}, { cspNonce, appDecorator, document }) {
```

**Comment by KasparRosin** [comment]
https://styled-components.com/docs/advanced#server-side-rendering
This is the documentation about SSR with styled-components. It's made to be used on api side, so i don't think it would make sense for us to bring it to client side logic as well.
I'm sure there are ways we can accomplish this, like by passing the sheet as a reference and do an additional check to see if we are dealing with SSR, but I don't really see how it makes the code more readable and understandable.

**Comment by allantatter** [comment]
@KasparRosin I think Matis's question was whether this can be accomplished like this:

```
    const { content, context } = await render(
      wrapperFunction(App),
      req.originalUrl,
      { config: getConfig() },
    );
```

Or does it need to wrap also `<AppDataContext.Provider>` and `<HelmetProvider>`?

**Comment by KasparRosin** [comment]
Oh sorry, misunderstood the question. Will test whether the suggestion works.

**Review by KasparRosin — APPROVED** [review]

**Comment by MatisLepik** [comment]
> @MatisLepik There were some issues with making a wrapper component such as we would have to re-implement the .getPageData functionality ourselves.

Ahh, fair enough. You might be able to do it with `CustomApp.getPageData = App.getPageData`, but it's already getting a bit hacky.

**Review by Tarpsvo — APPROVED** [review]

**Comment by Tarpsvo** [comment]
Looks good. Does require thorough upgrade guide and documentation though.

**Pull request merged**